compiler/src/iree/compiler/Preprocessing/Common/test/preprocessing_match_ops.mlir

```diff
@@ -271,3 +271,177 @@ module attributes {transform.with_named_sequence} {
+  transform.named_sequence @match_correct_maps(%op: !transform.any_op {transform.readonly}) -> !transform.any_op {
+   %batch, %m, %n, %k = transform.iree.match.is_contraction %op,
+    lhs_type = i8, rhs_type = i8, output_type = i32 {indexing_maps = [#map_matmul0, #map_matmul1, #map_matmul2]} :
+    (!transform.any_op) ->
+    (!transform.param<i64>, !transform.param<i64>, !transform.param<i64>, !transform.param<i64>)
+    transform.yield %op : !transform.any_op
+  }
+
+  transform.named_sequence @annotate_matched(%op: !transform.any_op {transform.readonly}) {
+    %0 = transform.param.constant "matched" -> !transform.any_param
+    transform.annotate %op "indexing_maps_match" = %0 : !transform.any_op, !transform.any_param
+    transform.yield
+  }
+
+  transform.named_sequence @__transform_main(%module: !transform.any_op) {
+    transform.foreach_match in %module
+        @match_correct_maps -> @annotate_matched
+      : (!transform.any_op) -> (!transform.any_op)
+    transform.yield
+  }
+}
+
+// -----
+
+// Verify that operations with different number of indexing maps are correctly not matched.
+
+#map_matmul0 = affine_map<(d0, d1, d2) -> (d0, d2)>
+#map_matmul1 = affine_map<(d0, d1, d2) -> (d1, d2)>
+#map_matmul2 = affine_map<(d0, d1, d2) -> (d0, d1)>
+
+// CHECK-LABEL: func.func @op_matmul
+func.func @op_matmul(%input0: tensor<32x64xi8>, %input1: tensor<32x64xi8>, %dest: tensor<32x32xi32>) -> tensor<32x32xi32> {
+  // CHECK-NEXT: linalg.matmul
+  // CHECK-SAME:   indexing_maps_match = "unmatched"
+  %res = linalg.matmul
+        indexing_maps = [#map_matmul0, #map_matmul1, #map_matmul2]
+        ins(%input0, %input1 : tensor<32x64xi8>, tensor<32x64xi8>)
+        outs(%dest : tensor<32x32xi32>) {indexing_maps_match = "unmatched"} -> tensor<32x32xi32>
+  return %res : tensor<32x32xi32>
+}
+
+module attributes {transform.with_named_sequence} {
+  transform.named_sequence @match_different_count(%op: !transform.any_op {transform.readonly}) -> !transform.any_op {
+    %batch, %m, %n, %k = transform.iree.match.is_contraction %op,
+      lhs_type = i8, rhs_type = i8, output_type = i32 {indexing_maps = [#map_matmul0, #map_matmul1, #map_matmul2, #map_matmul0]} :
+      (!transform.any_op) -> (!transform.param<i64>, !transform.param<i64>, !transform.param<i64>, !transform.param<i64>)
+    transform.yield %op : !transform.any_op
+  }
+
+  transform.named_sequence @annotate_matched(%op: !transform.any_op {transform.readonly}) {
+    %0 = transform.param.constant "matched" -> !transform.any_param
+    transform.annotate %op "indexing_maps_match" = %0 : !transform.any_op, !transform.any_param
+    transform.yield
+  }
+
+  transform.named_sequence @__transform_main(%module: !transform.any_op) {
+     // Should NOT match: operation has 3 indexing maps but matcher expects 4.
+    transform.foreach_match in %module
+        @match_different_count -> @annotate_matched
+      : (!transform.any_op) -> (!transform.any_op)
+    transform.yield
+  }
+}
```

Comment:
Can you add a testcase that shows if we match contractions that already have some attribute?